Docs/Book/Cookbook_v2.rst

```diff
@@ -0,0 +1,1009 @@
+
+| **Author:** Michael Palmer/ Greg Landrum
+| **Source:** `<https://sourceforge.net/p/rdkit/mailman/message/29679834/>`_
+| **Index ID#:** RDKitCB_11
+| **Summary:** Calculate the explicit valence, number of explicit and implicit hydrogens, and total number of hydrogens on an atom. See the link for an important explanation about terminology and implementation of these methods in RDKit. Highlights are presented below.
+
+Most of the time (exception is explained below), explicit refers to atoms that are in the graph and 
+implicit refers to atoms that are not in the graph (i.e., Hydrogens). So given that the ring is aromatic (e.g.,in pyrrole), 
+the explicit valence of each of the atoms (ignores the Hs that are not present in the graph) in pyrrole is 3. If you want the Hydrogen count,
+use GetTotalNumHs(); the total number of Hs for each atom is one:
+
+.. testcode::
+
+    from rdkit import Chem
+    pyrrole = Chem.MolFromSmiles('C1=CNC=C1')
+    for atom in pyrrole.GetAtoms():
+       print(atom.GetSymbol(), atom.GetExplicitValence(), atom.GetTotalNumHs())
+
+.. testoutput::
+
+   C 3 1
+   C 3 1
+   N 3 1
+   C 3 1
+   C 3 1
+
+In RDKit, there is overlapping nomenclature around the use of the words
+"explicit" and "implicit" when it comes to Hydrogens. When you specify the Hydrogens for an atom inside of square brackets 
+in the SMILES, it becomes an "explicit" hydrogen as far as atom.GetNumExplicitHs() is concerned. Here is an example:
+
+.. testcode::
+
+   pyrrole = Chem.MolFromSmiles('C1=CNC=C1')
+   mol1 = Chem.MolFromSmiles('C1=CNCC1')
+   mol2 = Chem.MolFromSmiles('C1=C[NH]CC1')
+
+.. testcode::
+
+   for atom in pyrrole.GetAtoms():
+       print(atom.GetSymbol(), atom.GetExplicitValence(), atom.GetNumImplicitHs(), atom.GetNumExplicitHs(), atom.GetTotalNumHs())
+
+.. testoutput::
+
+   C 3 1 0 1
+   C 3 1 0 1
+   N 3 0 1 1
+   C 3 1 0 1
+   C 3 1 0 1
+
+.. testcode::
+   
+    for atom in mol1.GetAtoms():
+       print(atom.GetSymbol(), atom.GetExplicitValence(), atom.GetNumImplicitHs(), atom.GetNumExplicitHs(), atom.GetTotalNumHs())
+
+.. testoutput::
+
+   C 3 1 0 1
+   C 3 1 0 1
+   N 2 1 0 1
+   C 2 2 0 2
+   C 2 2 0 2
+
+.. testcode::
+    
+    for atom in mol2.GetAtoms():
+       print(atom.GetSymbol(), atom.GetExplicitValence(), atom.GetNumImplicitHs(), atom.GetNumExplicitHs(), atom.GetTotalNumHs())
+
+.. testoutput::
+
+   C 3 1 0 1
+   C 3 1 0 1
+   N 3 0 1 1
+   C 2 2 0 2
+   C 2 2 0 2
+
+Wiener Index
+=============
+
+| **Author:** Greg Landrum
+| **Source:** `<https://sourceforge.net/p/rdkit/mailman/message/36802142/>`_
+| **Index ID#:** RDKitCB_12
+| **Summary:** Calculate the Wiener index (a topological index of a molecule)
+
+.. testcode::
+
+   from rdkit import Chem
+   def wiener_index(m):
+       res = 0
+       amat = Chem.GetDistanceMatrix(m)
+       for i in range(m.GetNumAtoms()):
+           for j in range(i+1,m.GetNumAtoms()):
+               res += amat[i][j]
+       return res
```

Comment:
Perhaps slightly clearer if you pull out `m.GetNumAtoms` into a separate variable?

Reply:
Thanks, corrected. I agree, that is a lot more clear now.